File: Python/excepciones/GatoSimple.py
'''
Created on 15 ene. 2019

@author: González Sabariego Francisco Javier
'''

from excepciones.ExcepcionApareamientoImposible import ExcepcionApareamientoImposible

class GatoSimple():
    
    def __init__(self, sexo):
        self.color="negro"
        self.raza="bombay"
        self.sexo=sexo
        self.edad=int(2)
        self.peso=float(4.1)
    
    #Dice el sexo del gato.
    def getSexo(self):
        return self.sexo
    
    #El gato maulla.
    def maulla(self):
        print("Miauuuuu")
    
    #El gato ronronea.
    def ronronea(self):
        print("Mrrrrrrr")
    
    #El gato come.
    def come(self, comida):
        if comida=="pescado":
            print("Hmmmm, gracias.")
        else:
            print("Lo siento, yo solo como pescado.")
    
    #El gato pelea con otro gato.
    def peleaCon(self, contrincante):
        if self.sexo=="hembra":
            print("No me gusta pelear.")
        else:
            if contrincante.getSexo()=="hembra":
                print("No me gusta pelear contra gatitas.")
            else:
                print("Ven aquí que te vas a enterar.")



    def apareaCon(self, gato):
        if self.getSexo()==gato.getSexo():
            raise ExcepcionApareamientoImposible()
        else:
            print("Acaba de nacer una cría.")


if __name__ == "__main__":
    
    #Gato Garfield
    garfield=GatoSimple("macho")
    
    print("hola gatito")
    garfield.maulla()
    print("toma tarta")
    garfield.come("tarta selva negra")
    print("toma pescado, a ver si esto te gusta")
    garfield.come("pescado")
    
    #Gato Tom
    tom=GatoSimple("macho")
    
    print("tom, toma sopita de verduras")
    tom.come("sopa de verduras")
    
    #Gata Lisa
    lisa=GatoSimple("hembra")
    print("gatitos a ver como maullais")
    garfield.maulla()
    tom.maulla()
    lisa.maulla()
    
    garfield.peleaCon(lisa)
    lisa.peleaCon(tom)
    tom.peleaCon(garfield)
    
    
    

File: Java/excepciones/GatoSimple.java
package excepciones;

public class GatoSimple {
  //##############   Atributos   ##############\\
  String color, raza, sexo;
  int edad;
  double peso;
  int numCrias = 0;
  
  //##############   Método   ##############\\
  
//constructor
  GatoSimple () {
    this.sexo = "";
  }
  
  //constructor
  GatoSimple (String s) {
    this.sexo = s;
  }
  
  
  //getter
  String getSexo() {
    return this.sexo;
  }
  
  
  /**
  * Hace que el gato maulle
  */
  void maulla() {
    System.out.println("Miauuuu");
  }
  
  
  /**
  * Hace que el gato ronronee
  */
  void ronronea() {
    System.out.println("mrrrrrr");
  }
  
  
  /**
  * Hace que el gato coma.
  * A los gatos les gusta el pescado, si le damos otra comida
  * la rechazará.
  *
  Programación orientada a objetos 123
  * @param comida la comida que se le ofrece al gato
  */
  void come(String comida) {
    if (comida.equals("pescado")) {
      System.out.println("Hmmmm, gracias");
    } else {
      System.out.println("Lo siento, yo solo como pescado");
    }
  }
  
  
  /**
  * Pone a pelear dos gatos.
  * Solo se van a pelear dos machos entre sí.
  *
  * @param contrincante es el gato contra el que pelear
  */
  void peleaCon(GatoSimple contrincante) {
    if (this.sexo.equals("hembra")) {
      System.out.println("No me gusta pelear");
    } else {
      if (contrincante.getSexo().equals("hembra")) {
        System.out.println("No peleo con gatitas");
      } else {
        System.out.println("Ven aquí que te vas a enterar");
      }
    }
  }
  
  
  /**
   * Permite aparearse a dos gatos de diferente sexo.
   * 
   * @param gato  Instancia de gato con el que el
   * @throws ExcepcionApareamientoImposible
   */
  void apareaCon(GatoSimple gato) throws ExcepcionApareamientoImposible{
    if(this.getSexo().equals(gato.getSexo())) {
      throw new ExcepcionApareamientoImposible();
    } else {
      System.out.println("Acaba de nacer una cría. Enhorabuena!!");
    }
  }
  
}
